src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java

```diff
@@ -213,92 +212,104 @@ public CassandraStorage(
+
+          if (4 <= currentVersion) {
+            List<String> otherRunningReapers = session.execute("SELECT reaper_instance_host FROM running_reapers").all()
+                .stream()
+                .map((row) -> row.getString("reaper_instance_host"))
+                .filter((reaperInstanceHost) -> !AppContext.REAPER_INSTANCE_ADDRESS.equals(reaperInstanceHost))
+                .collect(Collectors.toList());
+
+            Preconditions.checkState(
+                otherRunningReapers.isEmpty(),
+                "Database migration can not happen with other reaper instances running. Found ",
+                StringUtils.join(otherRunningReapers));
+          }
 
-    // initialize/upgrade db schema
-    try (Database database = new Database(cassandra, keyspace)) {
-      int currentVersion = database.getVersion();
-      MigrationRepository migrationRepo = new MigrationRepository("db/cassandra");
-      if (currentVersion < migrationRepo.getLatestVersion()) {
-        LOG.warn("Starting db migration from {} to {}…", currentVersion, migrationRepo.getLatestVersion());
-
-        if (4 <= currentVersion) {
-          List<String> otherRunningReapers = session.execute("SELECT reaper_instance_host FROM running_reapers").all()
-              .stream()
-              .map((row) -> row.getString("reaper_instance_host"))
-              .filter((reaperInstanceHost) -> !AppContext.REAPER_INSTANCE_ADDRESS.equals(reaperInstanceHost))
-              .collect(Collectors.toList());
-
-          Preconditions.checkState(
-              otherRunningReapers.isEmpty(),
-              "Database migration can not happen with other reaper instances running. Found ",
-              StringUtils.join(otherRunningReapers));
+          migrate(database.getVersion(), migrationRepo, session, schemaMigrationLock);
+          // some migration steps depend on the Cassandra version, so must be rerun every startup
+          Migration016.migrate(session, config.getCassandraFactory().getKeyspace());
+          // Switch metrics table to TWCS if possible, this is intentionally executed every startup
+          Migration021.migrate(session, config.getCassandraFactory().getKeyspace());
+        } else {
+          LOG.info(
+              String.format("Keyspace %s already at schema version %d", session.getLoggedKeyspace(), currentVersion));
         }
-
-        migrate(database.getVersion(), migrationRepo, session);
-        // some migration steps depend on the Cassandra version, so must be rerun every startup
-        Migration016.migrate(session, keyspace);
-        // Switch metrics table to TWCS if possible, this is intentionally executed every startup
-        Migration021.migrate(session, keyspace);
-      } else {
-        LOG.info(
-            String.format("Keyspace %s already at schema version %d", session.getLoggedKeyspace(), currentVersion));
+        schemaMigrationLock.unlockSchemaMigration();
```

Comment:
shouldn't this be inside a `} finally {` block?

Reply:
yes, putting it in a `finally` block

Reply:
i went further and used try-resources, see what i did in https://github.com/thelastpickle/cassandra-reaper/commit/eb7a757ba38be468f84c191197be2f9ba915ebfe